### PR TITLE
Fixes reduced delivery of Brave ads campaigns with subdivision targeting

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/18748

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See https://github.com/brave/brave-browser/issues/18748

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ads/inline_content_ad_exclusion_rules.cc [hunk 1]
```diff
@@ -36,98 +35,112 @@ ExclusionRules::ExclusionRules(
...
 
 ExclusionRules::~ExclusionRules() = default;
 
 bool ExclusionRules::ShouldExcludeCreativeAd(
-    const CreativeAdInfo& creative_ad) const {
-  bool should_exclude = false;
+    const CreativeAdInfo& creative_ad) {
+  if (WasExcluded(creative_ad)) {
+    return true;
+  }
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ads/inline_content_ad_exclusion_rules.cc:49]: Early return if creative ad was previously excluded

File: vendor/bat-native-ads/src/bat/ads/internal/ads/exclusion_rules_base.cc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    return true;
+  }
+
+  if (uuids_.find(creative_ad.campaign_id) != uuids_.end()) {
+    return true;
+  }
+
+  if (uuids_.find(creative_ad.advertiser_id) != uuids_.end()) {
+    return true;
+  }
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/exclusion_rules_base.cc:48]: Frequency caps exclude ads based on creative instance id, creative set id, campaign id or advertiser id

File: vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/ad_notifications/eligible_ad_notifications_v1.cc [hunk 3]
```diff
@@ -158,54 +158,31 @@ void EligibleAdsV1::GetForUntargeted(
...
   }
 
   CreativeAdNotificationList eligible_creative_ads = creative_ads;
 
+  frequency_capping::ExclusionRules exclusion_rules(
+      subdivision_targeting_, anti_targeting_resource_, ad_events,
+      browsing_history);
+  eligible_creative_ads = ApplyFrequencyCapping(
+      eligible_creative_ads, last_served_ad_, &exclusion_rules);
+
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/ad_notifications/eligible_ad_notifications_v1.cc:173]: Frequency capping must occur before filtering creative ads for seen ads/advertisers for round robin to work as expected

File: vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/round_robin_advertisers.h [hunk 4]
```diff
@@ -22,10 +21,8 @@ T FilterSeenAdvertisers(const T& ads,
   const auto iter = std::remove_if(
       unseen_advertisers.begin(), unseen_advertisers.end(),
       [&seen_advertisers](const CreativeAdInfo& creative_ad) {
-        SplitTestFrequencyCap frequency_cap;
-        return frequency_cap.ShouldExclude(creative_ad) ||
-               seen_advertisers.find(creative_ad.advertiser_id) !=
-                   seen_advertisers.end();
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/round_robin_advertisers.h:28]: Frequency capping now occurs before filtering creative ads for seen ads/advertisers, so this fix is no longer required.

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.h [hunk 5]
```diff
@@ -22,6 +22,8 @@ class ConversionFrequencyCap final : public ExclusionRule<CreativeAdInfo> {
   ConversionFrequencyCap(const ConversionFrequencyCap&) = delete;
   ConversionFrequencyCap& operator=(const ConversionFrequencyCap&) = delete;
 
+  std::string GetUuid(const CreativeAdInfo& creative_ad) const override;
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.h:25]: GetUUid should return the id for the frequency cap, i.e. creative instance id, creative set id, campaign id or advertiser id. This value should match the value logged in GetLastMessage for each frequency cap

File: vendor/bat-native-ads/src/bat/ads/internal/conversions/conversions.cc [hunk 6]
```diff
@@ -123,8 +123,9 @@ std::string ExtractConversionIdFromText(
   return conversion_id;
 }
 
-std::set<std::string> GetConvertedCreativeSets(const AdEventList& ad_events) {
-  std::set<std::string> creative_set_ids;
+std::unordered_set<std::string> GetConvertedCreativeSets(
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/conversions/conversions.cc:126]: changed to `std::unordered_set` to optimize performance

File: vendor/bat-native-ads/src/bat/ads/internal/locale/country_code_anonymity_set.h [hunk 7]
```diff
@@ -6,12 +6,12 @@
...
 #define BRAVE_VENDOR_BAT_NATIVE_ADS_SRC_BAT_ADS_INTERNAL_LOCALE_COUNTRY_CODE_ANONYMITY_SET_H_
 
-#include <set>
 #include <string>
+#include <unordered_set>
 
 namespace ads {
 
-const std::set<std::string> kCountryCodeAnonymitySet = {
+const std::unordered_set<std::string> kCountryCodeAnonymitySet = {
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/locale/country_code_anonymity_set.h:14]: changed to `std::unordered_set` to optimize performance

File: vendor/bat-native-ads/src/bat/ads/internal/locale/supported_country_codes_aliases.h [hunk 8]
```diff
@@ -7,12 +7,12 @@
...
 
 #include <map>
-#include <set>
 #include <string>
+#include <unordered_set>
 
 namespace ads {
 
-using SupportedCountryCodesSet = std::set<std::string>;
+using SupportedCountryCodesSet = std::unordered_set<std::string>;
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/locale/supported_country_codes_aliases.h:15]: changed to `std::unordered_set` to optimize performance

File: vendor/bat-native-ads/src/bat/ads/internal/segments/segments_util.cc [hunk 9]
```diff
@@ -32,7 +32,7 @@ std::vector<std::string> SplitSegment(const std::string& segment) {
 void RemoveDuplicates(SegmentList* segments) {
   DCHECK(segments);
 
-  std::set<std::string> seen;  // log(n) existence check
+  std::unordered_set<std::string> seen;  // log(n) existence check
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/segments/segments_util.cc:35]: changed to `std::unordered_set` to optimize performance

File: vendor/bat-native-ads/src/bat/ads/internal/ads/ad_notifications/ad_notification_exclusion_rules.cc [hunk 10]
```diff
@@ -36,103 +35,118 @@ ExclusionRules::ExclusionRules(
...
-  bool should_exclude = false;
+    const CreativeAdInfo& creative_ad) {
+  if (WasExcluded(creative_ad)) {
+    return true;
+  }
 
   DailyCapFrequencyCap daily_cap_frequency_cap(ad_events_);
   if (ShouldExclude(creative_ad, &daily_cap_frequency_cap)) {
-    should_exclude = true;
+    Exclude(daily_cap_frequency_cap.GetUuid(creative_ad));
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/ad_notifications/ad_notification_exclusion_rules.cc:53]: seems like `ShouldExclude` is not a proper name of this function now.
maybe we can rename function to somthing like `ExcludeIfNeeded` ?

File: vendor/bat-native-ads/src/bat/ads/internal/database/tables/ad_events_database_table.cc [hunk 11]
```diff
@@ -83,7 +83,7 @@ void AdEvents::PurgeExpired(ResultCallback callback) {
       "(SELECT creative_set_id from creative_ads) "
       "AND creative_set_id NOT IN "
       "(SELECT creative_set_id from creative_ad_conversions) "
-      "AND DATETIME('now') >= DATETIME(timestamp, 'unixepoch', '+3 month')",
+      "AND DATETIME('now') >= DATETIME(timestamp, 'unixepoch', '+1 month')",
```
[the remark this answers is not in the document]
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/database/tables/ad_events_database_table.cc:86]: Do we have database index for this field?

File: vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/ad_notifications/eligible_ad_notifications_v2.cc [hunk 12]
```diff
@@ -115,26 +86,19 @@ void EligibleAdsV2::GetEligibleAds(
   });
 }
 
-CreativeAdNotificationList EligibleAdsV2::ApplyFrequencyCapping(
+CreativeAdNotificationList EligibleAdsV2::FilterCreativeAds(
     const CreativeAdNotificationList& creative_ads,
-    const AdInfo& last_served_ad,
     const AdEventList& ad_events,
-    const BrowsingHistoryList& browsing_history) const {
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/eligible_ads/ad_notifications/eligible_ad_notifications_v2.cc:122]: Why this method became non const?

File: vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ads/inline_content_ad_exclusion_rules.cc [hunk 13]
```diff
@@ -36,98 +35,112 @@ ExclusionRules::ExclusionRules(
...
 
   DailyCapFrequencyCap daily_cap_frequency_cap(ad_events_);
   if (ShouldExclude(creative_ad, &daily_cap_frequency_cap)) {
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ads/inline_content_ad_exclusion_rules.cc:52]: Fixed

**Pull request merged**